From 99fddfab520986a3d1d49e4f5d80ba4b0eabaf04 Mon Sep 17 00:00:00 2001
From: Nikita Chernikov <nchernikov@maxlinear.com>
Date: Wed, 20 May 2026 11:16:51 +0300
Subject: [PATCH] WLANRTSYS-102001 Fix non-TX BSS critical update not set post
 ML CSA

Fixup to: WLANRTSYS-100621 MLO triband critical update changes

Signed-off-by: Nikita Chernikov <nchernikov@maxlinear.com>
---
 src/ap/drv_callbacks.c |  8 ++++----
 src/ap/mxl_mld.c       | 31 +++++++++++++++++++++++++++++++
 src/ap/mxl_mld.h       |  1 +
 3 files changed, 36 insertions(+), 4 deletions(-)

diff --git a/src/ap/drv_callbacks.c b/src/ap/drv_callbacks.c
index 8cfa5de89..1ce86c1e4 100644
--- a/src/ap/drv_callbacks.c
+++ b/src/ap/drv_callbacks.c
@@ -1304,10 +1304,6 @@ void hostapd_event_ch_switch(struct hostapd_data *hapd, int freq, int ht,
 		       width, channel_width_to_string(width), cf1, cf2,
 		       punct_bitmap);
 
-#ifdef CONFIG_IEEE80211BE_MXL_MLO
-	if (mxl_is_mlo_enabled(hapd->conf))
-		mxl_ml_clear_csa_params_affected_link(hapd);
-#endif
 
 	if (!hapd->iface->current_mode) {
 		hostapd_logger(hapd, NULL, HOSTAPD_MODULE_IEEE80211,
@@ -1544,6 +1540,10 @@ void hostapd_event_ch_switch(struct hostapd_data *hapd, int freq, int ht,
 	/* No need to set other radio beacons after the CSA is done. We have proprietary
 	   changes to handle RNR updates both in case of 6GHz colocation and MLO */
 	ieee802_11_set_beacon_per_bss_only(hapd);
+#ifdef CONFIG_IEEE80211BE_MXL_MLO
+	if (mxl_is_mlo_enabled(hapd->conf))
+		mxl_ml_handle_post_csa(hapd);
+#endif /* CONFIG_IEEE80211BE_MXL_MLO */
 #endif
 
 #ifdef CONFIG_VENDOR_MXL
diff --git a/src/ap/mxl_mld.c b/src/ap/mxl_mld.c
index 3f8226d87..c22222f1d 100644
--- a/src/ap/mxl_mld.c
+++ b/src/ap/mxl_mld.c
@@ -3238,6 +3238,37 @@ static int mxl_ml_eht_op_change_critical_update(struct hostapd_iface *csa_iface)
 	return 0;
 }
 
+int mxl_ml_handle_post_csa(struct hostapd_data *hapd)
+{
+	struct hostapd_iface *csa_iface = hapd->iface;
+	struct hostapd_data *bss, *tx_bss;
+	struct p_ap_mld *mld = NULL;
+	int i;
+
+	mld = mxl_ml_get_ap_mld(hapd);
+	if (!mld)
+		return 0;
+
+	if (csa_iface->conf->multibss_enable) {
+		tx_bss = mxl_hostapd_mbssid_get_tx_bss(csa_iface->bss[0]);
+		/* Handle only for TX BSS */
+		if (hapd != tx_bss)
+			return 0;
+		/* Cleanup all non-TX BSS only after TX BSS beacon was updated */
+		for (i = 0; i < csa_iface->num_bss; i++) {
+			bss = csa_iface->bss[i];
+			mld = mxl_ml_get_ap_mld(bss);
+			if (!mld)
+				continue;
+			mxl_ml_clear_csa_params_affected_link(bss);
+		}
+	} else {
+		mxl_ml_clear_csa_params_affected_link(hapd);
+	}
+
+	return 0;
+}
+
 static void mxl_ml_csa_timeout_handler(void *eloop_data,
 							   void *user_ctx)
 {
diff --git a/src/ap/mxl_mld.h b/src/ap/mxl_mld.h
index fb613beef..de532f372 100644
--- a/src/ap/mxl_mld.h
+++ b/src/ap/mxl_mld.h
@@ -330,6 +330,7 @@ int mxl_ml_send_bss_critical_update_info(struct hostapd_data *hapd,
 int mxl_ml_handle_critical_update(struct hostapd_data *hapd);
 int mxl_ml_handle_csa_critical_update(struct hostapd_iface *csa_iface,
 					   struct csa_settings *settings, u8 op_class);
+int mxl_ml_handle_post_csa(struct hostapd_data *hapd);
 void mxl_cleanup_ml_csa_params(struct hostapd_iface *iface, bool cu_cleanup);
 void mxl_ml_clear_csa_params_affected_link(struct hostapd_data *hapd);
 void mxl_ml_clear_csa_params_reporting_link(struct hostapd_data *hapd);
-- 
2.43.0

